# Datasets

Teste planejados:
  1. Teste com diferentes dataset (Somente normalizado, Nomalizado e oversampling e normalizado e under sampling)
  2. Teste de hiperparametros

Qual vai ser a metodologia? Primeiro vou testar os 3 datasets com a condições default para ter a baseline e depois vou testar os 3 com os melhores parametros achados pelo GridSearch 2 vezes (2 vezes buscando os melhores resultados)

In [9]:
import pandas as pd

C:\Users\mathe\AppData\Local\Temp\ipykernel_5700\4080736814.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [10]:
df_norm = pd.read_csv('../datasets/data_normalizada.csv')
df_norm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6819 entries, 0 to 6818
Data columns (total 94 columns):
 #   Column                                                   Non-Null Count  Dtype  
---  ------                                                   --------------  -----  
 0   Bankrupt?                                                6819 non-null   int64  
 1   ROA(C) before interest and depreciation before interest  6819 non-null   float64
 2   ROA(A) before interest and % after tax                   6819 non-null   float64
 3   ROA(B) before interest and depreciation after tax        6819 non-null   float64
 4   Operating Gross Margin                                   6819 non-null   float64
 5   Realized Sales Gross Margin                              6819 non-null   float64
 6   Operating Profit Rate                                    6819 non-null   float64
 7   Pre-tax net Interest Rate                                6819 non-null   float64
 8   After-tax net Interest Rate 

In [11]:
df_over = pd.read_csv('../datasets/data_smote.csv')
df_over.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7924 entries, 0 to 7923
Data columns (total 94 columns):
 #   Column                                                   Non-Null Count  Dtype  
---  ------                                                   --------------  -----  
 0   Bankrupt?                                                7924 non-null   int64  
 1   ROA(C) before interest and depreciation before interest  7924 non-null   float64
 2   ROA(A) before interest and % after tax                   7924 non-null   float64
 3   ROA(B) before interest and depreciation after tax        7924 non-null   float64
 4   Operating Gross Margin                                   7924 non-null   float64
 5   Realized Sales Gross Margin                              7924 non-null   float64
 6   Operating Profit Rate                                    7924 non-null   float64
 7   Pre-tax net Interest Rate                                7924 non-null   float64
 8   After-tax net Interest Rate 

In [12]:
df_under = pd.read_csv('../datasets/data_under.csv')
df_under.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 258 entries, 0 to 257
Data columns (total 94 columns):
 #   Column                                                   Non-Null Count  Dtype  
---  ------                                                   --------------  -----  
 0   Bankrupt?                                                258 non-null    int64  
 1   ROA(C) before interest and depreciation before interest  258 non-null    float64
 2   ROA(A) before interest and % after tax                   258 non-null    float64
 3   ROA(B) before interest and depreciation after tax        258 non-null    float64
 4   Operating Gross Margin                                   258 non-null    float64
 5   Realized Sales Gross Margin                              258 non-null    float64
 6   Operating Profit Rate                                    258 non-null    float64
 7   Pre-tax net Interest Rate                                258 non-null    float64
 8   After-tax net Interest Rate   

Dividindo entre treino e teste

# Baselines

In [13]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
from sklearn.ensemble import VotingClassifier
from sklearn.model_selection import train_test_split

def resultados(nome = ''):
  dataset_atual = []
  if nome == '':
    dataset_atual = df_norm
  elif nome == 'over':
    dataset_atual = df_over
  elif nome == 'under':
    dataset_atual = df_under

  y = dataset_atual['Bankrupt?']

  # Removendo a coluna alvo do DataFrame original para obter apenas as features
  X = dataset_atual.drop(columns=['Bankrupt?'])

  # Dividindo os dados em conjuntos de treinamento e teste
  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

 
  # Decision Tree
  dt_classifier = DecisionTreeClassifier()
  dt_classifier.fit(X_train, y_train)

  # Random Forest
  rf_classifier = RandomForestClassifier()
  rf_classifier.fit(X_train, y_train)

  # Decision Tree
  y_pred_dt = dt_classifier.predict(X_test)

  # Random Forest
  y_pred_rf = rf_classifier.predict(X_test)

  # Avaliação do Decision Tree
  print(f"Relatório de Classificação para Decision Tree {nome}:")
  print(classification_report(y_test, y_pred_dt))

  print('')
  # Avaliação do Random Forest
  print(f"Relatório de Classificação para Random Forest {nome}:")
  print(classification_report(y_test, y_pred_rf))

  print('')

## Dataset normalizado

Usando os modelos nas condições default

In [14]:
resultados()

Relatório de Classificação para Decision Tree :
              precision    recall  f1-score   support

           0       0.98      0.97      0.97      1313
           1       0.33      0.37      0.35        51

    accuracy                           0.95      1364
   macro avg       0.65      0.67      0.66      1364
weighted avg       0.95      0.95      0.95      1364


Relatório de Classificação para Random Forest :
              precision    recall  f1-score   support

           0       0.97      1.00      0.98      1313
           1       0.67      0.12      0.20        51

    accuracy                           0.96      1364
   macro avg       0.82      0.56      0.59      1364
weighted avg       0.96      0.96      0.95      1364




Dá pra notar que a classe 1 não está sendo muito bem reconhecida!

## Dataset Over Sampling

In [16]:
resultados(nome="over")

Relatório de Classificação para Decision Tree over:
              precision    recall  f1-score   support

           0       0.98      0.93      0.96       798
           1       0.94      0.98      0.96       787

    accuracy                           0.96      1585
   macro avg       0.96      0.96      0.96      1585
weighted avg       0.96      0.96      0.96      1585


Relatório de Classificação para Random Forest over:
              precision    recall  f1-score   support

           0       1.00      0.97      0.98       798
           1       0.97      1.00      0.98       787

    accuracy                           0.98      1585
   macro avg       0.98      0.98      0.98      1585
weighted avg       0.98      0.98      0.98      1585




Com oversampling os resultados ficaram muito mais positivos! Com o comitê sendo o melhor e o KNN sendo o pior

## Dataset Under Sampling

In [15]:
resultados(nome="under")

Relatório de Classificação para Decision Tree under:
              precision    recall  f1-score   support

           0       0.83      0.83      0.83        29
           1       0.78      0.78      0.78        23

    accuracy                           0.81        52
   macro avg       0.81      0.81      0.81        52
weighted avg       0.81      0.81      0.81        52


Relatório de Classificação para Random Forest under:
              precision    recall  f1-score   support

           0       0.85      0.79      0.82        29
           1       0.76      0.83      0.79        23

    accuracy                           0.81        52
   macro avg       0.81      0.81      0.81        52
weighted avg       0.81      0.81      0.81        52




Os resultados ficaram mais positivos do que o dataset somente normalizado. Apesar disso, ficou um pouco pior do que o oversampling.

# Testes de parametros

## Definindo a 1a grade de procura para o GridSearch

In [25]:
parameters_dt = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 5, 10, 20],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

parameters_rf = {
    'n_estimators': [100, 200, 300],
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 5, 10, 20],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

## Decision Tree

In [26]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report

def busca_resultados_dt(nome='', parameters=0):
    dataset_atual = ''
    max_depth_passada = None
    if nome == '':
        dataset_atual = df_norm
        if parameters:
            max_depth_passada = 5
    elif nome == 'over':
        dataset_atual = df_over
        if parameters:
            max_depth_passada = 10
    elif nome == 'under':
        dataset_atual = df_under
        if parameters:
            max_depth_passada = 5

    y = dataset_atual['Bankrupt?']
    X = dataset_atual.drop(columns=['Bankrupt?'])

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    dt = DecisionTreeClassifier()
    parametersAtual = parameters_dt

    if parameters:
        parametersAtual = parameters_dt_2
        dt = DecisionTreeClassifier(max_depth=max_depth_passada)

    grid_search = GridSearchCV(estimator=dt, param_grid=parametersAtual, cv=3)
    grid_search.fit(X_train, y_train)

    best_params = grid_search.best_params_
    print("Melhores parâmetros encontrados:", best_params)

    best_model = grid_search.best_estimator_
    y_pred = best_model.predict(X_test)
    report = classification_report(y_test, y_pred)
    print("Relatório de Classificação do melhor modelo:")
    print(report)


## Random Forest

In [21]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report
def busca_resultados_rf(nome='', parameters=0):
  dataset_atual = ''
  max_depth_passada = None
  if nome == '':
    dataset_atual = df_norm
    if parameters:
      max_depth_passada = 5
  elif nome == 'over':
    dataset_atual = df_over
    if parameters:
      max_depth_passada = 10
  elif nome == 'under':
    dataset_atual = df_under
    if parameters:
      max_depth_passada = 5

  y = dataset_atual['Bankrupt?']
  X = dataset_atual.drop(columns=['Bankrupt?'])

  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

  rf = RandomForestClassifier()
  parametersAtual = parameters_rf

  if parameters:
    parametersAtual = parameters_rf_2
    rf = RandomForestClassifier(max_depth=max_depth_passada)

  grid_search = GridSearchCV(estimator=rf, param_grid=parametersAtual, cv=3)
  grid_search.fit(X_train, y_train)

  best_params = grid_search.best_params_
  print("Melhores parâmetros encontrados:", best_params)

  best_model = grid_search.best_estimator_
  y_pred = best_model.predict(X_test)
  report = classification_report(y_test, y_pred)
  print("Relatório de Classificação do melhor modelo:")
  print(report)

## Resultados para as primeiras grades

### Decision Tree

In [27]:
print("Base somente normalizada")
busca_resultados_dt()
print("")
print("Base com oversampling")
busca_resultados_dt(nome="over")
print("")
print("Base com undersampling")
busca_resultados_dt(nome="under")

Base somente normalizada
Melhores parâmetros encontrados: {'criterion': 'gini', 'max_depth': 5, 'min_samples_leaf': 2, 'min_samples_split': 2}
Relatório de Classificação do melhor modelo:
              precision    recall  f1-score   support

           0       0.97      0.99      0.98      1313
           1       0.45      0.25      0.33        51

    accuracy                           0.96      1364
   macro avg       0.71      0.62      0.65      1364
weighted avg       0.95      0.96      0.96      1364


Base com oversampling
Melhores parâmetros encontrados: {'criterion': 'entropy', 'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2}
Relatório de Classificação do melhor modelo:
              precision    recall  f1-score   support

           0       0.97      0.94      0.96       798
           1       0.94      0.97      0.96       787

    accuracy                           0.96      1585
   macro avg       0.96      0.96      0.96      1585
weighted avg       0.

### Random Forest

In [28]:
print("Base somente normalizada")
busca_resultados_rf()
print("")
print("Base com oversampling")
busca_resultados_rf(nome="over")
print("")
print("Base com undersampling")
busca_resultados_rf(nome="under")

Base somente normalizada
Melhores parâmetros encontrados: {'criterion': 'gini', 'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 10, 'n_estimators': 100}
Relatório de Classificação do melhor modelo:
              precision    recall  f1-score   support

           0       0.97      1.00      0.98      1313
           1       1.00      0.18      0.30        51

    accuracy                           0.97      1364
   macro avg       0.98      0.59      0.64      1364
weighted avg       0.97      0.97      0.96      1364


Base com oversampling
Melhores parâmetros encontrados: {'criterion': 'gini', 'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 300}
Relatório de Classificação do melhor modelo:
              precision    recall  f1-score   support

           0       0.99      0.96      0.98       798
           1       0.96      0.99      0.98       787

    accuracy                           0.98      1585
   macro avg       0.98      0.9

## Definindo a 2a grade para procurar

In [31]:
parameters_dt_2 = {
    'splitter': ['best', 'random'],
    'max_leaf_nodes': [None, 10, 20, 30, 40],
    'min_impurity_decrease': [0.0, 0.1, 0.2],
    'class_weight': [None, 'balanced']
}

parameters_rf_2 = {
    'max_samples': [None, 0.5, 0.7, 0.9],
    'max_leaf_nodes': [None, 10, 20, 30, 40],
    'min_impurity_decrease': [0.0, 0.1, 0.2],
    'class_weight': [None, 'balanced', 'balanced_subsample'],
    'ccp_alpha': [0.0, 0.1, 0.2]
}

## Resultados para as segundas grades

### Decision Tree

In [32]:
print("Base somente normalizada")
busca_resultados_dt(parameters=1)
print("")
print("Base com oversampling")
busca_resultados_dt(nome="over", parameters=1)
print("")
print("Base com undersampling")
busca_resultados_dt(nome="under", parameters=1)

Base somente normalizada
Melhores parâmetros encontrados: {'class_weight': None, 'max_leaf_nodes': None, 'min_impurity_decrease': 0.1, 'splitter': 'best'}
Relatório de Classificação do melhor modelo:
              precision    recall  f1-score   support

           0       0.96      1.00      0.98      1313
           1       0.00      0.00      0.00        51

    accuracy                           0.96      1364
   macro avg       0.48      0.50      0.49      1364
weighted avg       0.93      0.96      0.94      1364


Base com oversampling


c:\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Melhores parâmetros encontrados: {'class_weight': None, 'max_leaf_nodes': None, 'min_impurity_decrease': 0.0, 'splitter': 'best'}
Relatório de Classificação do melhor modelo:
              precision    recall  f1-score   support

           0       0.98      0.91      0.94       798
           1       0.92      0.98      0.95       787

    accuracy                           0.95      1585
   macro avg       0.95      0.95      0.95      1585
weighted avg       0.95      0.95      0.95      1585


Base com undersampling
Melhores parâmetros encontrados: {'class_weight': 'balanced', 'max_leaf_nodes': 30, 'min_impurity_decrease': 0.0, 'splitter': 'random'}
Relatório de Classificação do melhor modelo:
              precision    recall  f1-score   support

           0       0.80      0.83      0.81        29
           1       0.77      0.74      0.76        23

    accuracy                           0.79        52
   macro avg       0.79      0.78      0.78        52
weighted avg       0.

### Random Forest

In [33]:
print("Base somente normalizada")
busca_resultados_rf(parameters=1)
print("")
print("Base com oversampling")
busca_resultados_rf(nome="over", parameters=1)
print("")
print("Base com undersampling")
busca_resultados_rf(nome="under", parameters=1)

Base somente normalizada
Melhores parâmetros encontrados: {'ccp_alpha': 0.0, 'class_weight': None, 'max_leaf_nodes': 40, 'max_samples': 0.5, 'min_impurity_decrease': 0.0}
Relatório de Classificação do melhor modelo:
              precision    recall  f1-score   support

           0       0.97      1.00      0.98      1313
           1       1.00      0.10      0.18        51

    accuracy                           0.97      1364
   macro avg       0.98      0.55      0.58      1364
weighted avg       0.97      0.97      0.95      1364


Base com oversampling
Melhores parâmetros encontrados: {'ccp_alpha': 0.0, 'class_weight': None, 'max_leaf_nodes': None, 'max_samples': None, 'min_impurity_decrease': 0.0}
Relatório de Classificação do melhor modelo:
              precision    recall  f1-score   support

           0       1.00      0.94      0.97       798
           1       0.94      1.00      0.97       787

    accuracy                           0.97      1585
   macro avg       0.9

# Resultados

Vale notar que todos os resultados obtidos após a busca de parâmetros de cada modelo, em todos os datasets, foram melhores do que a baseline

## Decision Tree

### Base de dados normalizada

Nessas condições o melhor resultado foi obtido usando os parâmetros obtidos pelo 1º GridSearch, que deu o seguinte resultado:


- Melhores parâmetros encontrados: {'criterion': 'gini', 'max_depth': 5, 'min_samples_leaf': 2, 'min_samples_split': 2}

Relatório de Classificação do melhor modelo:
              
                  precision    recall  f1-score   support

               0       0.97      0.99      0.98      1313
               1       0.45      0.25      0.33        51

        accuracy                           0.96      1364
       macro avg       0.71      0.62      0.65      1364
    weighted avg       0.95      0.96      0.96      1364    

### Base de dados com oversampling

Nesse database o melhor resultado foi encontrado também pelo primeiro GridSearch feito. Os resultados foram os seguintes:


- Melhores parâmetros para o comitê de redes neurais: {'criterion': 'entropy', 'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2}

Relatório de Classificação do melhor modelo do comitê de redes neurais:
  
                    precision    recall  f1-score   support

               0       0.97      0.94      0.96       798
               1       0.94      0.97      0.96       787

        accuracy                           0.96      1585
       macro avg       0.96      0.96      0.96      1585
    weighted avg       0.96      0.96      0.96      1585    

### Base de dados com undersampling

Nesse caso o melhor conjunto de parâmetros foi encontrado no 1º GridSearch. O resultado foi o seguinte:


- Melhores parâmetros encontrados: {'criterion': 'gini', 'max_depth': 5, 'min_samples_leaf': 1, 'min_samples_split': 5}

Relatório de Classificação do melhor modelo:
              
                    precision    recall  f1-score   support

               0       0.84      0.93      0.89        29
               1       0.90      0.78      0.84        23

        accuracy                           0.87        52
       macro avg       0.87      0.86      0.86        52
    weighted avg       0.87      0.87      0.86        52

## Random Forest

### Base de dados normalizada

Nesse contexto o melhor resultado foi encontrado pelo 1º GridSearch, com a seguinte configuração:

- Melhores parâmetros encontrados: {'criterion': 'gini', 'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 10, 'n_estimators': 100}

Relatório de Classificação do melhor modelo:
              
                    precision    recall  f1-score   support

               0       0.97      1.00      0.98      1313
               1       1.00      0.18      0.30        51

        accuracy                           0.97      1364
       macro avg       0.98      0.59      0.64      1364
    weighted avg       0.97      0.97      0.96      1364     

### Base de dados com oversampling

Nesse caso o melhor resultado foi obtido com os parâmetros encontrados no 1º GridSearch com a seguinte configuração:

- Melhores parâmetros encontrados: {'criterion': 'gini', 'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 300}

Relatório de Classificação do melhor modelo:
              
                    precision    recall  f1-score   support

               0       0.99      0.96      0.98       798
               1       0.96      0.99      0.98       787

        accuracy                           0.98      1585
       macro avg       0.98      0.98      0.98      1585
    weighted avg       0.98      0.98      0.98      1585    

### Base de dados com undersampling

Nesse caso o melhor resultado foi obtido com o 2º GridSearch com a seguinte configuração (vale notar que o 1º atingiu o mesmo resultado):


- Melhores parâmetros encontrados: {'ccp_alpha': 0.0, 'class_weight': 'balanced', 'max_leaf_nodes': 30, 'max_samples': None, 'min_impurity_decrease': 0.0}

Relatório de Classificação do melhor modelo:

                    precision    recall  f1-score   support

               0       0.88      0.79      0.84        29
               1       0.77      0.87      0.82        23

        accuracy                           0.83        52
       macro avg       0.83      0.83      0.83        52
    weighted avg       0.83      0.83      0.83        52     

## Extrair regras da decision tree

In [42]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report
from sklearn import tree

nome='' 
dataset_atual = ''
max_depth_passada = None
dataset_atual = df_norm

y = dataset_atual['Bankrupt?']
X = dataset_atual.drop(columns=['Bankrupt?'])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create the Decision Tree classifier
dt = DecisionTreeClassifier(criterion='gini', max_depth=None, min_samples_leaf=1, min_samples_split=2)

# Fit the classifier to the data
dt.fit(X_train, y_train)

# Get the feature importances
importances = dt.feature_importances_

# Get the indices of the most important features
top_indices = importances.argsort()[:10]  # Change the number 3 to the desired number of top features

# Get the names of the most important features
top_features = X.columns[top_indices]

# Print the first attributes used in the construction of the decision tree
print("First attributes used in the construction of the decision tree:")
for feature in top_features:
    print(feature)
    
# Predict the labels of the test set
y_pred = dt.predict(X_test)

text_representation = tree.export_text(dt, feature_names= X.columns.tolist())
print(text_representation)


First attributes used in the construction of the decision tree:
Average Collection Days
Cash Reinvestment %
Current Ratio
Quick Ratio
Total debt/Total net worth
Long-term fund suitability ratio (A)
Total Asset Turnover
Accounts Receivable Turnover
Net Worth Turnover Rate (times)
Revenue per person
|--- Net Income to Stockholder's Equity <= -0.42
|   |--- Interest Expense Ratio <= -0.07
|   |   |--- Non-industry income and expenditure/revenue <= -0.02
|   |   |   |--- Current Liability to Liability <= -1.20
|   |   |   |   |--- Gross Profit to Sales <= -1.53
|   |   |   |   |   |--- class: 1
|   |   |   |   |--- Gross Profit to Sales >  -1.53
|   |   |   |   |   |--- class: 0
|   |   |   |--- Current Liability to Liability >  -1.20
|   |   |   |   |--- Total Asset Return Growth Rate Ratio <= -0.19
|   |   |   |   |   |--- Cash/Total Assets <= -0.83
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- Cash/Total Assets >  -0.83
|   |   |   |   |   |   |--- class: 0
|   |   |   

10 primeiros atributos usados na construção da árvore de decisão:

First attributes used in the construction of the decision tree:
Average Collection Days
Cash Reinvestment %
Current Ratio
Quick Ratio
Total debt/Total net worth
Long-term fund suitability ratio (A)
Total Asset Turnover
Accounts Receivable Turnover
Net Worth Turnover Rate (times)
Revenue per person

Regras da árvore de decisão: 

In [ ]:
|--- Net Income to Stockholder's Equity <= -0.42
|   |--- Interest Expense Ratio <= -0.07
|   |   |--- Non-industry income and expenditure/revenue <= -0.02
|   |   |   |--- Current Liability to Liability <= -1.20
|   |   |   |   |--- Gross Profit to Sales <= -1.53
|   |   |   |   |   |--- class: 1
|   |   |   |   |--- Gross Profit to Sales >  -1.53
|   |   |   |   |   |--- class: 0
|   |   |   |--- Current Liability to Liability >  -1.20
|   |   |   |   |--- Total Asset Return Growth Rate Ratio <= -0.19
|   |   |   |   |   |--- Cash/Total Assets <= -0.83
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- Cash/Total Assets >  -0.83
|   |   |   |   |   |   |--- class: 0
|   |   |   |   |--- Total Asset Return Growth Rate Ratio >  -0.19
|   |   |   |   |   |--- Inventory/Working Capital <= -0.19
|   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |--- Inventory/Working Capital >  -0.19
|   |   |   |   |   |   |--- class: 1
|   |   |--- Non-industry income and expenditure/revenue >  -0.02
|   |   |   |--- Operating profit per person <= -0.91
|   |   |   |   |--- Cash Flow to Sales <= 0.00
|   |   |   |   |   |--- class: 1
|   |   |   |   |--- Cash Flow to Sales >  0.00
|   |   |   |   |   |--- class: 0
|   |   |   |--- Operating profit per person >  -0.91
|   |   |   |   |--- class: 0
|   |--- Interest Expense Ratio >  -0.07
|   |   |--- Net Income to Total Assets <= -4.94
|   |   |   |--- Inventory and accounts receivable/Net value <= 0.38
|   |   |   |   |--- Net Value Per Share (C) <= -1.42
|   |   |   |   |   |--- class: 0
|   |   |   |   |--- Net Value Per Share (C) >  -1.42
|   |   |   |   |   |--- Total income/Total expense <= -0.05
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- Total income/Total expense >  -0.05
|   |   |   |   |   |   |--- class: 0
|   |   |   |--- Inventory and accounts receivable/Net value >  0.38
|   |   |   |   |--- class: 1
|   |   |--- Net Income to Total Assets >  -4.94
|   |   |   |--- Research and development expense rate <= 1.09
|   |   |   |   |--- Cash Flow to Total Assets <= -0.16
|   |   |   |   |   |--- Operating Profit Growth Rate <= -0.02
|   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |--- Operating Profit Growth Rate >  -0.02
|   |   |   |   |   |   |--- Debt ratio % <= 0.11
|   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |--- Debt ratio % >  0.11
|   |   |   |   |   |   |   |--- Current Liability to Assets <= 1.88
|   |   |   |   |   |   |   |   |--- Total Asset Return Growth Rate Ratio <= -0.23
|   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |--- Total Asset Return Growth Rate Ratio >  -0.23
|   |   |   |   |   |   |   |   |   |--- Inventory/Working Capital <= 0.06
|   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |   |--- Inventory/Working Capital >  0.06
|   |   |   |   |   |   |   |   |   |   |--- Operating Profit Rate <= -0.00
|   |   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |   |   |--- Operating Profit Rate >  -0.00
|   |   |   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |--- Current Liability to Assets >  1.88
|   |   |   |   |   |   |   |   |--- Quick Assets/Total Assets <= 1.17
|   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |--- Quick Assets/Total Assets >  1.17
|   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |--- Cash Flow to Total Assets >  -0.16
|   |   |   |   |   |--- Current Asset Turnover Rate <= 2.90
|   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |--- Current Asset Turnover Rate >  2.90
|   |   |   |   |   |   |--- class: 1
|   |   |   |--- Research and development expense rate >  1.09
|   |   |   |   |--- Net Income to Total Assets <= -3.48
|   |   |   |   |   |--- class: 1
|   |   |   |   |--- Net Income to Total Assets >  -3.48
|   |   |   |   |   |--- Equity to Long-term Liability <= 0.68
|   |   |   |   |   |   |--- Retained Earnings to Total Assets <= -2.38
|   |   |   |   |   |   |   |--- Current Liability to Current Assets <= -0.46
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |--- Current Liability to Current Assets >  -0.46
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |--- Retained Earnings to Total Assets >  -2.38
|   |   |   |   |   |   |   |--- Net Value Per Share (C) <= -0.75
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |--- Net Value Per Share (C) >  -0.75
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- Equity to Long-term Liability >  0.68
|   |   |   |   |   |   |--- Net Income to Stockholder's Equity <= -1.39
|   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |--- Net Income to Stockholder's Equity >  -1.39
|   |   |   |   |   |   |   |--- class: 1
|--- Net Income to Stockholder's Equity >  -0.42
|   |--- Borrowing dependency <= 0.28
|   |   |--- Net Income to Total Assets <= -2.81
|   |   |   |--- Regular Net Profit Growth Rate <= 0.01
|   |   |   |   |--- class: 1
|   |   |   |--- Regular Net Profit Growth Rate >  0.01
|   |   |   |   |--- class: 0
|   |   |--- Net Income to Total Assets >  -2.81
|   |   |   |--- Continuous Net Profit Growth Rate <= -3.69
|   |   |   |   |--- class: 1
|   |   |   |--- Continuous Net Profit Growth Rate >  -3.69
|   |   |   |   |--- Current Assets/Total Assets <= 2.16
|   |   |   |   |   |--- Degree of Financial Leverage (DFL) <= -0.06
|   |   |   |   |   |   |--- ROA(B) before interest and depreciation after tax <= -1.26
|   |   |   |   |   |   |   |--- Total income/Total expense <= -0.04
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |--- Total income/Total expense >  -0.04
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |--- ROA(B) before interest and depreciation after tax >  -1.26
|   |   |   |   |   |   |   |--- Total Asset Return Growth Rate Ratio <= 0.17
|   |   |   |   |   |   |   |   |--- Degree of Financial Leverage (DFL) <= -0.06
|   |   |   |   |   |   |   |   |   |--- Operating Profit Growth Rate <= -0.53
|   |   |   |   |   |   |   |   |   |   |--- Cash Flow to Sales <= 0.00
|   |   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |   |   |--- Cash Flow to Sales >  0.00
|   |   |   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |   |--- Operating Profit Growth Rate >  -0.53
|   |   |   |   |   |   |   |   |   |   |--- Working Capital/Equity <= 0.41
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 6
|   |   |   |   |   |   |   |   |   |   |--- Working Capital/Equity >  0.41
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 3
|   |   |   |   |   |   |   |   |--- Degree of Financial Leverage (DFL) >  -0.06
|   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |--- Total Asset Return Growth Rate Ratio >  0.17
|   |   |   |   |   |   |   |   |--- Retained Earnings to Total Assets <= -0.89
|   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |--- Retained Earnings to Total Assets >  -0.89
|   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- Degree of Financial Leverage (DFL) >  -0.06
|   |   |   |   |   |   |--- Cash/Total Assets <= -0.89
|   |   |   |   |   |   |   |--- Operating Profit Growth Rate <= 0.01
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |--- Operating Profit Growth Rate >  0.01
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |--- Cash/Total Assets >  -0.89
|   |   |   |   |   |   |   |--- Net worth/Assets <= 2.08
|   |   |   |   |   |   |   |   |--- Degree of Financial Leverage (DFL) <= 11.05
|   |   |   |   |   |   |   |   |   |--- Interest Expense Ratio <= 0.12
|   |   |   |   |   |   |   |   |   |   |--- Research and development expense rate <= 3.04
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 4
|   |   |   |   |   |   |   |   |   |   |--- Research and development expense rate >  3.04
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 2
|   |   |   |   |   |   |   |   |   |--- Interest Expense Ratio >  0.12
|   |   |   |   |   |   |   |   |   |   |--- Interest Expense Ratio <= 0.12
|   |   |   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |   |   |--- Interest Expense Ratio >  0.12
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 5
|   |   |   |   |   |   |   |   |--- Degree of Financial Leverage (DFL) >  11.05
|   |   |   |   |   |   |   |   |   |--- Quick Asset Turnover Rate <= 1.23
|   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |   |--- Quick Asset Turnover Rate >  1.23
|   |   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |--- Net worth/Assets >  2.08
|   |   |   |   |   |   |   |   |--- Net Value Per Share (A) <= -0.99
|   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |--- Net Value Per Share (A) >  -0.99
|   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |--- Current Assets/Total Assets >  2.16
|   |   |   |   |   |--- Current Liabilities/Equity <= 2.40
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- Current Liabilities/Equity >  2.40
|   |   |   |   |   |   |--- class: 0
|   |--- Borrowing dependency >  0.28
|   |   |--- Non-industry income and expenditure/revenue <= -0.02
|   |   |   |--- Interest Expense Ratio <= -0.32
|   |   |   |   |--- ROA(B) before interest and depreciation after tax <= -0.55
|   |   |   |   |   |--- Cash flow rate <= -0.23
|   |   |   |   |   |   |--- Non-industry income and expenditure/revenue <= -0.07
|   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |--- Non-industry income and expenditure/revenue >  -0.07
|   |   |   |   |   |   |   |--- Current Liability to Current Assets <= 0.00
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |--- Current Liability to Current Assets >  0.00
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- Cash flow rate >  -0.23
|   |   |   |   |   |   |--- class: 0
|   |   |   |   |--- ROA(B) before interest and depreciation after tax >  -0.55
|   |   |   |   |   |--- Gross Profit to Sales <= -0.56
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- Gross Profit to Sales >  -0.56
|   |   |   |   |   |   |--- class: 0
|   |   |   |--- Interest Expense Ratio >  -0.32
|   |   |   |   |--- Cash Flow to Sales <= 0.00
|   |   |   |   |   |--- Fixed Assets Turnover Frequency <= 2.57
|   |   |   |   |   |   |--- Research and development expense rate <= 0.28
|   |   |   |   |   |   |   |--- Non-industry income and expenditure/revenue <= -0.02
|   |   |   |   |   |   |   |   |--- Persistent EPS in the Last Four Seasons <= -0.99
|   |   |   |   |   |   |   |   |   |--- Interest Coverage Ratio (Interest expense to EBIT) <= -0.09
|   |   |   |   |   |   |   |   |   |   |--- Current Liability to Equity <= 0.07
|   |   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |   |   |--- Current Liability to Equity >  0.07
|   |   |   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |   |--- Interest Coverage Ratio (Interest expense to EBIT) >  -0.09
|   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |--- Persistent EPS in the Last Four Seasons >  -0.99
|   |   |   |   |   |   |   |   |   |--- Total Asset Growth Rate <= 0.62
|   |   |   |   |   |   |   |   |   |   |--- Non-industry income and expenditure/revenue <= -0.02
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 2
|   |   |   |   |   |   |   |   |   |   |--- Non-industry income and expenditure/revenue >  -0.02
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 2
|   |   |   |   |   |   |   |   |   |--- Total Asset Growth Rate >  0.62
|   |   |   |   |   |   |   |   |   |   |--- Current Liability to Assets <= 0.95
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 2
|   |   |   |   |   |   |   |   |   |   |--- Current Liability to Assets >  0.95
|   |   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |--- Non-industry income and expenditure/revenue >  -0.02
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |--- Research and development expense rate >  0.28
|   |   |   |   |   |   |   |--- Net Value Per Share (A) <= -0.43
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |--- Net Value Per Share (A) >  -0.43
|   |   |   |   |   |   |   |   |--- Inventory Turnover Rate (times) <= -0.62
|   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |--- Inventory Turnover Rate (times) >  -0.62
|   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- Fixed Assets Turnover Frequency >  2.57
|   |   |   |   |   |   |--- Realized Sales Gross Margin <= -0.20
|   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |--- Realized Sales Gross Margin >  -0.20
|   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |--- Cash Flow to Sales >  0.00
|   |   |   |   |   |--- Inventory and accounts receivable/Net value <= -0.59
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- Inventory and accounts receivable/Net value >  -0.59
|   |   |   |   |   |   |--- ROA(C) before interest and depreciation before interest <= -1.38
|   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |--- ROA(C) before interest and depreciation before interest >  -1.38
|   |   |   |   |   |   |   |--- Total Asset Return Growth Rate Ratio <= 0.14
|   |   |   |   |   |   |   |   |--- Current Asset Turnover Rate <= 2.67
|   |   |   |   |   |   |   |   |   |--- Net profit before tax/Paid-in capital <= -1.32
|   |   |   |   |   |   |   |   |   |   |--- Current Liabilities/Liability <= -1.79
|   |   |   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |   |   |--- Current Liabilities/Liability >  -1.79
|   |   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |   |--- Net profit before tax/Paid-in capital >  -1.32
|   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |--- Current Asset Turnover Rate >  2.67
|   |   |   |   |   |   |   |   |   |--- Total income/Total expense <= -0.02
|   |   |   |   |   |   |   |   |   |   |--- Cash Flow Per Share <= -0.73
|   |   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |   |   |--- Cash Flow Per Share >  -0.73
|   |   |   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |   |--- Total income/Total expense >  -0.02
|   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |--- Total Asset Return Growth Rate Ratio >  0.14
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |--- Non-industry income and expenditure/revenue >  -0.02
|   |   |   |--- Working Capital/Equity <= -1.44
|   |   |   |   |--- class: 1
|   |   |   |--- Working Capital/Equity >  -1.44
|   |   |   |   |--- Fixed Assets Turnover Frequency <= 3.38
|   |   |   |   |   |--- Persistent EPS in the Last Four Seasons <= -0.57
|   |   |   |   |   |   |--- Continuous Net Profit Growth Rate <= -0.00
|   |   |   |   |   |   |   |--- Persistent EPS in the Last Four Seasons <= -0.58
|   |   |   |   |   |   |   |   |--- Net Value Per Share (B) <= -1.34
|   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |--- Net Value Per Share (B) >  -1.34
|   |   |   |   |   |   |   |   |   |--- Contingent liabilities/Net worth <= 0.10
|   |   |   |   |   |   |   |   |   |   |--- Cash/Total Assets <= -0.86
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 2
|   |   |   |   |   |   |   |   |   |   |--- Cash/Total Assets >  -0.86
|   |   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |   |--- Contingent liabilities/Net worth >  0.10
|   |   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |--- Persistent EPS in the Last Four Seasons >  -0.58
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |--- Continuous Net Profit Growth Rate >  -0.00
|   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- Persistent EPS in the Last Four Seasons >  -0.57
|   |   |   |   |   |   |--- Operating Funds to Liability <= 0.20
|   |   |   |   |   |   |   |--- Degree of Financial Leverage (DFL) <= 0.90
|   |   |   |   |   |   |   |   |--- Equity to Long-term Liability <= 0.84
|   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |--- Equity to Long-term Liability >  0.84
|   |   |   |   |   |   |   |   |   |--- Equity to Long-term Liability <= 0.86
|   |   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |   |--- Equity to Long-term Liability >  0.86
|   |   |   |   |   |   |   |   |   |   |--- Borrowing dependency <= 0.90
|   |   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |   |   |--- Borrowing dependency >  0.90
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 2
|   |   |   |   |   |   |   |--- Degree of Financial Leverage (DFL) >  0.90
|   |   |   |   |   |   |   |   |--- Net profit before tax/Paid-in capital <= -0.44
|   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |--- Net profit before tax/Paid-in capital >  -0.44
|   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |--- Operating Funds to Liability >  0.20
|   |   |   |   |   |   |   |--- ROA(C) before interest and depreciation before interest <= 0.79
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |--- ROA(C) before interest and depreciation before interest >  0.79
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |--- Fixed Assets Turnover Frequency >  3.38
|   |   |   |   |   |--- Current Liability to Liability <= -0.63
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- Current Liability to Liability >  -0.63
|   |   |   |   |   |   |--- class: 0



De acordo com a regra abaixo, se uma entrada tiver  'Net Income to Stockholder's Equity <= -0.42', 'Interest Expense Ratio <= -0.07', 'Non-industry income and expenditure/revenue <= -0.02' e 'Current Liability to Liability <= -1.20', mas tiver 'Gross Profit to Sales <= -1.53', então será da classe 1, bankrupt. 

Porém, se tiver os anteriores e 'Gross Profit to Sales >  -1.53', será da classe 0, não bankrupt.

In [ ]:
|--- Net Income to Stockholder's Equity <= -0.42
|   |--- Interest Expense Ratio <= -0.07
|   |   |--- Non-industry income and expenditure/revenue <= -0.02
|   |   |   |--- Current Liability to Liability <= -1.20
|   |   |   |   |--- Gross Profit to Sales <= -1.53
|   |   |   |   |   |--- class: 1
|   |   |   |   |--- Gross Profit to Sales >  -1.53
|   |   |   |   |   |--- class: 0